File: go/collections_test.go
package tachyon

import (
	"context"
	"net/http"
	"testing"
)

func TestCollectionsCreate(t *testing.T) {
	client, requests := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 201, `{"name":"products","fields":[{"name":"title","type":"text"}],"num_documents":0,"num_segments":0}`)
	})

	info, err := client.Collections.Create(context.Background(), CollectionSchema{
		Name:   "products",
		Fields: []FieldSchema{{Name: "title", Type: FieldTypeText}},
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if info.Name != "products" || info.NumDocuments != 0 {
		t.Fatalf("unexpected info: %+v", info)
	}

	if len(*requests) != 1 {
		t.Fatalf("expected 1 request, got %d", len(*requests))
	}
	req := (*requests)[0]
	if req.Method != http.MethodPost || req.Path != "/collections" {
		t.Fatalf("unexpected request: %+v", req)
	}
	if req.Header.Get("X-TACHYON-API-KEY") != "admin-key" {
		t.Fatalf("missing api key header: %+v", req.Header)
	}
}

func TestCollectionsList(t *testing.T) {
	client, _ := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, `[{"name":"products","fields":[],"num_documents":5,"num_segments":1}]`)
	})

	list, err := client.Collections.List(context.Background())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(list) != 1 || list[0].Name != "products" || list[0].NumDocuments != 5 {
		t.Fatalf("unexpected list: %+v", list)
	}
}

func TestCollectionsRetrieve(t *testing.T) {
	client, requests := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, `{"name":"products","fields":[],"num_documents":5,"num_segments":1}`)
	})

	info, err := client.Collections.Retrieve(context.Background(), "products")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if info.NumDocuments != 5 {
		t.Fatalf("unexpected info: %+v", info)
	}
	if (*requests)[0].Path != "/collections/products" {
		t.Fatalf("unexpected path: %s", (*requests)[0].Path)
	}
}

func TestCollectionsRetrieveURLEncodesName(t *testing.T) {
	client, requests := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, `{"name":"my products","fields":[],"num_documents":0,"num_segments":0}`)
	})

	_, err := client.Collections.Retrieve(context.Background(), "my products")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if (*requests)[0].Path != "/collections/my%20products" {
		t.Fatalf("unexpected path: %s", (*requests)[0].Path)
	}
}

func TestCollectionsDelete(t *testing.T) {
	client, requests := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(204)
	})

	if err := client.Collections.Delete(context.Background(), "products"); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if (*requests)[0].Method != http.MethodDelete || (*requests)[0].Path != "/collections/products" {
		t.Fatalf("unexpected request: %+v", (*requests)[0])
	}
}

File: java/src/test/java/io/github/adikeshri/tachyon/CollectionsTest.java
package io.github.adikeshri.tachyon;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class CollectionsTest {

    @Test
    void create_sendsSchemaAndReturnsInfo() throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.json(201,
            """
            {"name":"products","fields":[{"name":"title","type":"text"}],"num_documents":0,"num_segments":0}
            """))) {
            TachyonClient client = TestClients.forServer(server);

            CollectionInfo info = client.collections.create(CollectionSchema.builder("products")
                .fields(FieldSchema.builder("title", FieldType.TEXT).build())
                .build());

            assertEquals("products", info.name());
            assertEquals(0, info.numDocuments());
            assertEquals(1, server.requests.size());
            TestServer.CapturedRequest request = server.requests.get(0);
            assertEquals("POST", request.method());
            assertEquals("/collections", request.path());
            assertEquals("admin-key", request.header("X-TACHYON-API-KEY"));
            assertTrue(request.body().contains("\"title\""));
        }
    }

    @Test
    void list_returnsCollections() throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.json(200,
            """
            [{"name":"products","fields":[],"num_documents":5,"num_segments":1}]
            """))) {
            TachyonClient client = TestClients.forServer(server);

            List<CollectionInfo> list = client.collections.list();

            assertEquals(1, list.size());
            assertEquals("products", list.get(0).name());
            assertEquals(5, list.get(0).numDocuments());
        }
    }

    @Test
    void retrieve_returnsCollection() throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.json(200,
            """
            {"name":"products","fields":[],"num_documents":5,"num_segments":1}
            """))) {
            TachyonClient client = TestClients.forServer(server);

            CollectionInfo info = client.collections.retrieve("products");

            assertEquals(5, info.numDocuments());
            assertEquals("/collections/products", server.requests.get(0).path());
        }
    }

    @Test
    void retrieve_urlEncodesCollectionName() throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.json(200,
            """
            {"name":"my products","fields":[],"num_documents":0,"num_segments":0}
            """))) {
            TachyonClient client = TestClients.forServer(server);

            client.collections.retrieve("my products");

            assertEquals("/collections/my%20products", server.requests.get(0).path());
        }
    }

    @Test
    void delete_sendsDeleteAndCompletesOn204() throws IOException {
        try (TestServer server = new TestServer(req -> TestServer.FakeResponse.noContent())) {
            TachyonClient client = TestClients.forServer(server);

            assertDoesNotThrow(() -> client.collections.delete("products"));

            assertEquals("DELETE", server.requests.get(0).method());
            assertEquals("/collections/products", server.requests.get(0).path());
        }
    }
}
